troubleshooting/known_issues/kafka_chart_upgrade_failure.md

```diff
@@ -0,0 +1,214 @@
+Name: "kafkarebalances.kafka.strimzi.io", Namespace: ""
+for: "/tmp/strimzi-crds-0.41.0.yaml": CustomResourceDefinition.apiextensions.k8s.io "kafkarebalances.kafka.strimzi.io" is invalid: status.storedVersions[1]: Invalid value: "v1alpha1": must appear in spec.versions
+```
+
+Similar type of errors could be seen on other Strimzi CRDs too.
+
+## Workaround steps
+
+(`ncn-mw#`) To resolve this issue, update the Strimzi CRDs on the cluster by removing references to the older versions - `v1alpha1` and `v1beta1`.
+After this update, the CRDs will retain only the latest supported version `v1beta2` which eliminates the upgrade failure.
+
+Before applying the workaround, we need access to update the CRDs. For that, run the following command which would create ServiceAccount, Role and RoleBinding -
+
+```bash
+kubectl apply -f - <<EOF
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: crd-updater
+  namespace: default
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crd-update-access
+rules:
+  - apiGroups: ["apiextensions.k8s.io"]
+    resources: ["customresourcedefinitions"]
+    verbs: ["get", "list", "watch","update","patch"]
+  - apiGroups: ["apiextensions.k8s.io"]
+    resources: ["customresourcedefinitions/status"]
+    verbs: ["get", "patch", "update"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: crd-update-binding
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: crd-update-access
+subjects:
+  - kind: ServiceAccount
+    name: crd-updater
+    namespace: default
+EOF
+```
+
+Now, run the following script to perform the workaround -
+
+```bash
+strimzi_kafka_CRD_list=$(kubectl get crd -l app=strimzi -o custom-columns=":metadata.name")
+VERSION_TO_KEEP="v1beta2"
+
+# Grab the Kubernetes API server URL from your current kubeconfig
+APISERVER=$(kubectl config view --minify -o jsonpath='{.clusters[0].cluster.server}')
+echo "Using API Server: $APISERVER"
+
+# Get the service account token from crd-updater
+SECRET_NAME=$(kubectl get sa crd-updater -n default -o jsonpath='{.secrets[0].name}')
+TOKEN=$(kubectl get secret "$SECRET_NAME" -n default -o jsonpath='{.data.token}' | base64 -d)
+
+for CRD in "${strimzi_kafka_CRD_list[@]}"; do
+  echo "Processing CRD: $CRD"
+
+  CRD_JSON=$(kubectl get crd "$CRD" -o json)
+
+  #Update the "served" and "storage" fields to true for the version v1beta2 and also Remove all versions except v1beta2 from spec.versions
+  MODIFIED_SPEC=$(echo "$CRD_JSON" | jq '
+    (.spec.versions[] | select(.name == "v1beta2")) |= (.served = true | .storage = true)
+    | .spec.versions |= map(select(.name == "v1beta2"))
+  ')
+
+  # Replace the CRD spec
+  echo "$MODIFIED_SPEC" | kubectl replace --raw "/apis/apiextensions.k8s.io/v1/customresourcedefinitions/$CRD" -f -
+
+  # Patch status.storedVersions to only include v1beta2
```

Comment:
Patching of `status.storedVersions` needs to be done before replacing the CRD spec.